provider/server/exec.go

```diff
@@ -5,11 +5,165 @@ package server
 
 import (
 	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
 
+	"github.com/creack/pty"
+	"github.com/digitalocean/go-libvirt"
+	"github.com/go-logr/logr"
 	iri "github.com/ironcore-dev/ironcore/iri/apis/machine/v1alpha1"
+	remotecommandserver "github.com/ironcore-dev/ironcore/poollet/machinepoollet/iri/streaming/remotecommand"
+	"github.com/ironcore-dev/libvirt-provider/pkg/api"
+	libvirtutils "github.com/ironcore-dev/libvirt-provider/pkg/libvirt/utils"
+	"github.com/moby/term"
+	"k8s.io/client-go/tools/remotecommand"
 )
 
+const (
+	StreamCreationTimeout = 30 * time.Second
+	StreamIdleTimeout     = 2 * time.Minute
+)
+
+type executorExec struct {
+	Libvirt         *libvirt.Libvirt
+	ExecRequest     *iri.ExecRequest
+	VirshExecutable string
+	Machine         *api.Machine
+}
+
 func (s *Server) Exec(ctx context.Context, req *iri.ExecRequest) (*iri.ExecResponse, error) {
+	machineID := req.MachineId
+	log := s.loggerFrom(ctx, "MachineID", machineID)
+
+	log.V(1).Info("Inserting request into cache")
+	token, err := s.execRequestCache.Insert(req)
+	if err != nil {
+		return nil, err
+	}
+
+	log.V(1).Info("Returning url with token")
+	return &iri.ExecResponse{
+		Url: s.buildURL("exec", token),
+	}, nil
+}
+
+func (s *Server) ServeExec(w http.ResponseWriter, req *http.Request, token string) {
```

Comment:
The 'unhappy' path of `ServeExec` should be also testet. Once we spin up a `VM` in the pipeline we can do the entire flow.

Reply:
The meaningful tests for `ServeExec` could be a little complex to mock, and better handled in the pipeline, considering its nature and dependencies. Nevertheless, I have added a simple test in the http-server for the flow of unknown tokens, that eventually calls ServeExec.